Api category

hi0826 opened the pull request:
### 📕 Issue Number

Close #90 
<br/>

### 📙 작업 내역

> 카테고리 관련 API

- [x] 카테고리 추가
- [x] 카테고리 수정
- [x] 카테고리 삭제
- 내역, 결제수단, 카테고리 관련 API는 동작이 동일합니다.
1. 해당하는 가계부를 탐색
2. 가계부에 있는 결제수단 멤버에 접근
3. 결제수단을 수정
4. 가계부를 update
<br/>

### 멘토님 멘셔닝

@boostcamp-2020/accountbook_mentor
<br/>

hi0826 committed:
[feat] #51 내역 모델, 인터페이스 변경

- 날짜를 date 타입이아닌 string으로 받게 변경

hi0826 committed:
[feat] #46 내역추가 api 변경

- 내역추가 api accountbook service에서 transaction service로 이동

hi0826 committed:
[feat] #46 내역 controller

- 내역 controller 기능추가
- patch는 추후 더 구현

hi0826 committed:
[feat] accountbook schema 변경

- 스키마가 좀 더 직관적으로 보이게 변경
- static 함수 추가중

hi0826 committed:
[feat] #50 내역 수정 api 구현중

- 내역 수정 api 구현중

hi0826 committed:
[feat] #50 가계부 내역 수정 API

- 내역 수정시 api 기능구현

hi0826 committed:
Merge branch 'develop' of https://github.com/boostcamp-2020/Project16-C-Account-Book into api-transaction

hi0826 committed:
[fix] #51 가계부 내역 모델 수정

- 가계부 내역 모델 Doc 오타 수정

hi0826 committed:
[feat] #50 내역 수정 API 완성

- 내역 수정 API 완성
- updateInfo에 _id를 넣어주지않았더니 _id가 바뀌는 현상 수정
- response 객체에 맞게 잘 전달

hi0826 committed:
[feat] #48 내역 삭제 API, 내역 수정 API 코드스타일 수정

- 내역 삭제 API 구현
- 내역 수정 API 코드스타일 수정

hi0826 committed:
[fix] transactions serice 수정

- Context 모듈 받아오는 부분 수정

hi0826 committed:
[feat] #53 결제수단 추가 API

- 결제수단 추가 API
- 내역 추가 API와 동일
- AccountBook에서 찾은 후 AccountBook을 update 하는 방식

hi0826 committed:
[feat] #55 결제수단 수정 API 추가

- 가계부별 결제수단 수정 API 추가

hi0826 committed:
[feat] #54 결제수단 삭제 API, 결제수단 수정 API

- 결제수단 삭제 API 구현
- 결제수단 수정 API 오타 수정

hi0826 committed:
[style] 결제수단 절대경로 수정

- 코드 스타일 수정

hi0826 committed:
[feat] #90 카테고리 추가, 수정 API

- 카테고리 추가 API
- 카테고리 수정 API

hi0826 committed:
[feat] #90 카테고리 삭제 API

- 카테고리 삭제 API 구현

j03y14 commented on BE/src/models/accountbook/index.ts:
```diff
@@ -77,19 +75,197 @@ const addTransaction = async (
-    const curTransactions = [...curAccountBook.transactions];
+    const curTransactions = curAccountBook.transactions;
     curTransactions.push(transaction);
     const updateResult = await AccountBookModel.update(
-      { _id: id },
-      { transactions: curTransactions },
+      { _id: accountBookId },
+      { transactions: [...curTransactions, transaction] },
+    );
```
이러면 transaction을 push 하고 또 [...curTransactions, transaction] 하니까 두번씩 들어가는거 아닌가요?

hi0826 replied:
어라 실수가있었네요 바로 고칠게요!

j03y14 commented on BE/src/models/accountbook/index.ts:
```diff
@@ -77,19 +75,197 @@ const addTransaction = async (
+): Promise<any> => {
+  const curAccountBook = await AccountBookModel.findOne({ _id: accountBookId });
+  if (curAccountBook) {
+    const index = curAccountBook.transactions
+      .map(value => value._id)
+      .indexOf(transactionId);
+    curAccountBook.transactions.splice(index, 1);
```
filter로 해도 됐을 것 같네요!
```typescript
const transactions = curAccountBook.transactions.filter(value => value._id !== transactionId);
```

hi0826 replied:
그생각을못했네요 ㅎㅅㅎ

j03y14 reviewed (approved):
수고하셨습니다!

hi0826 committed:
[fix] #46 내역추가 api 버그수정

- 코드 오류 찾아서 수정

hi0826 committed:
[feat] #8 #20 #38 가계부 모델 수정

- 가계부 모델에 startday(시작요일) 추가
  - startday는 숫자로 들어간다.
  - 0 1 2 3 4 5 6 ( 일~토 ) 순
  - 시작요일 변경은 가계부 상세 정보와 상관 없으므로, create와 update가 간단

hi0826 committed:
[feat] #8 #91 카테고리 모델 변경

- 카테고리에 type 추가
  - 수입/지출 을 카테고리 속성에 추가
-service 동작에 type 추가

yoonwoo123 commented:
깔끔하게 잘 짜주셨네요, 감사합니다!
